probe-rs/src/architecture/arm/sequences/cc13xx_cc26xx.rs

```diff
@@ -0,0 +1,355 @@
+        state: &mut JtagState,
+    ) -> Result<(), ArmError> {
+        // Build the DR value based on the requested operation. The DR value
+        // is based on the input arguments and contains several bitfields
+        let dr: u32 = ((rw & 0x1) << 31)
+            | ((block & 0x7) << 28)
+            | ((register & 0xF) << 24)
+            | (payload & 0xFFFFFF);
+
+        // Load IR with the router instruction, this allows us to write or read a data register
+        self.shift_ir(interface, IR_ROUTER, state, JtagState::SelectDRScan)?;
+        // Load the data register with the register block, address, and data
+        // The address and value to be written is encoded in the DR value
+        self.shift_dr(interface, 32, dr as u64, state, JtagState::SelectDRScan)?;
+        Ok(())
+    }
+
+    /// Does setup of the ICEPICK
+    ///
+    /// This will setup the ICEPICK to have the CPU/DAP on the scan chain and
+    /// also power and enable the debug interface for use with probe-rs
+    ///
+    /// This is a direct port of the openocd implementation:
+    /// https://github.com/openocd-org/openocd/blob/master/tcl/target/icepick.cfg#L81-L124
+    /// A few things were removed to fit the cc13xx_cc26xx family.
+    fn enable_icepick(
+        &self,
+        interface: &mut dyn DapProbe,
+        state: &mut JtagState,
+    ) -> Result<(), ArmError> {
+        let port = 0;
+        let block = 0x02;
+
+        // Select the Connect register
+        self.shift_ir(interface, IR_CONNECT, state, JtagState::SelectDRScan)?;
+        // Enable write, set the `ConnectKey` to 0b1001 (0x9) as per TRM section 6.3.3
+        self.shift_dr(interface, 8, 0x89, state, JtagState::SelectDRScan)?;
+        // Write to register 1 in the ICEPICK control block - keep JTAG powered in test logic reset
+        self.icepick_router(interface, 1, 0, 1, 0x000080, state)?;
+        // Write to register 0 in the Debug TAP linking block (Section 6.3.4.3)
+        // Namely:
+        // * [20]   : `InhibitSleep`
+        // * [16:14]: `ResetControl == Wait In Reset`
+        // * [8]    : `SelectTAP == 1`
+        // * [3]    : `ForceActive == Enable clocks`
+        self.icepick_router(interface, 1, block, port, 0x110108, state)?;
+
+        // Enter the bypass state
+        self.shift_ir(interface, IR_BYPASS, state, JtagState::RunTestIdle)?;
+
+        // Remain in run-test idle for 10 cycles
+        interface.jtag_sequence(10, false, set_n_bits(10))?;
+
+        Ok(())
+    }
+}
+
+impl ArmDebugSequence for CC13xxCC26xx {
+
+    fn reset_system(
+        &self,
+        _interface: &mut dyn ArmProbe,
+        _core_type: crate::CoreType,
+        _debug_base: Option<u64>,
+    ) -> Result<(), ArmError> {
+
+        // The only reset that is supported is system reset, which will reset the entire chip
+        // losing the JTAG configuration (back in two pin JTAG mode)
+        // TODO! Find a way to reset setup JTAG again after reset
+        Ok(())
```

Comment:
@Yatekii, I need some help here. This device converts CPU resets to system resets which undoes all the JTAG setup we did in `debug_port_setup`.

I overrode this temporarily so I could test my flash algorithm, but now I have basically render reset useless.

Is there a way that I can invoke `debug_port_setup` again?

Reply:
Hmm good question. I think @Dirbaio had the same problem.

Reply:
Is this a bug? Because I think ARM cores should be able to do soft resets (default impl of this sequence).

Reply:
Datasheet for the reference: https://www.ti.com/lit/ug/swcu185f/swcu185f.pdf?ts=1695616231827

Reply:
> Is this a bug? Because I think ARM cores should be able to do soft resets (default impl of this sequence).

It's certainly odd, and I believe this is supported, just that the soft reset is converted to a hard reset.

I can check with some of the folks our TI's tools team to double check how its done in their IDE.

Reply:
So indeed all MCU resets are converted to system reset which will reset the JTAG and require setup again.

There is a different register that does a warm reset for this device.

Reply:
Can we use that register? Should be possible to use it in the sequence, no?

Reply:
Yep, that's what I do now. It works.

I need to get my hands on some of the CC26x0 to be sure the sequence I coded up there works (of course its slightly different), but its working for the x2 variants.

Reply:
- [ ] Test with x0 and x4 variants

Reply:
I think we can assume that it works and if it does not we can open an issue and get to fix it. Except if you have some targets at hand. Incomplete functionality (1/3 chips) is better than none.